placement/algo/helper.go

```diff
@@ -0,0 +1,373 @@
+	return newPlaceShardingHelper(ps, s.Replicas(), false), nil
+}
+
+func newRemoveHostPlacementHelper(s placement.Snapshot, leavingHost placement.Host) (PlacementHelper, placement.HostShards, error) {
+	if s.HostShard(leavingHost.ID()) == nil {
+		return nil, nil, errHostAbsent
+	}
+	var leavingHostShards placement.HostShards
+	var hosts []placement.HostShards
+	for _, phs := range s.HostShards() {
+		if phs.Host().ID() == leavingHost.ID() {
+			leavingHostShards = phs
+			continue
+		}
+		hosts = append(hosts, phs)
+	}
+	ps := placement.NewPlacementSnapshot(hosts, s.Shards(), s.Replicas())
+	return newPlaceShardingHelper(ps, s.Replicas(), true), leavingHostShards, nil
+}
+
+func newPlaceShardingHelper(ps placement.Snapshot, targetRF int, hostCapacityAscending bool) PlacementHelper {
+	ph := &placementHelper{
+		shardToHostMap: make(map[uint32]map[placement.HostShards]struct{}),
+		rackToHostsMap: make(map[string]map[placement.HostShards]struct{}),
+		rf:             targetRF,
+		hostShards:     ps.HostShards(),
+		uniqueShards:   ps.Shards(),
+	}
+
+	// build rackToHost map
+	ph.buildRackToHostMap()
+
+	ph.buildHostHeap(hostCapacityAscending)
+	return ph
+}
+
+func (ph *placementHelper) buildRackToHostMap() {
+	for _, h := range ph.hostShards {
+		if _, exist := ph.rackToHostsMap[h.Host().Rack()]; !exist {
+			ph.rackToHostsMap[h.Host().Rack()] = make(map[placement.HostShards]struct{})
+		}
+		ph.rackToHostsMap[h.Host().Rack()][h] = struct{}{}
+		for _, shard := range h.Shards() {
+			ph.assignShardToHost(shard, h)
+		}
+	}
+}
+
+func (ph *placementHelper) buildHostHeap(hostCapacityAscending bool) {
+	overSizedRack := 0
+	overSizedHosts := 0
+	rackSizeMap := make(map[string]int)
+	for rack := range ph.rackToHostsMap {
+		rackHostNumber := len(ph.rackToHostsMap[rack])
+		if float64(rackHostNumber)/float64(ph.getHostLen()) >= 1.0/float64(ph.rf) {
+			overSizedRack++
+			overSizedHosts += rackHostNumber
+		}
+		rackSizeMap[rack] = rackHostNumber
+	}
+
+	targetLoadMap := ph.buildTargetLoadMap(rackSizeMap, overSizedRack, overSizedHosts)
+	ph.targetLoad = targetLoadMap
+	ph.hostHeap = newHostHeap(ph.hostShards, hostCapacityAscending, targetLoadMap, ph.rackToHostsMap)
+}
+
+func (ph *placementHelper) buildTargetLoadMap(rackSizeMap map[string]int, overSizedRackLen, overSizedHostLen int) map[string]int {
+	targetLoad := make(map[string]int)
+	for _, host := range ph.hostShards {
+		rackSize := rackSizeMap[host.Host().Rack()]
+		if float64(rackSize)/float64(ph.getHostLen()) >= 1.0/float64(ph.rf) {
+			// if the host is on a over-sized rack, the target load is topped at shardLen / rackSize
+			targetLoad[host.Host().ID()] = int(math.Ceil(float64(ph.getShardLen()) / float64(rackSize)))
+		} else {
+			// if the host is on a normal rack, get the target load with aware of other over-sized rack
+			targetLoad[host.Host().ID()] = ph.getShardLen() * (ph.rf - overSizedRackLen) / (ph.getHostLen() - overSizedHostLen)
```

Comment:
what happens if there are more hosts than `rf * number of shards`? Should we do a `math.Ceil` here?